Updating Jupyter docs with feedback from summer school in Heilbronn (#172)
eth-cscs/cscs-docs: docs/access/jupyterlab.md

Once logged in, you will be redirected to the JupyterHub Spawner Options form, where typical job configuration options can be selected in order to allocate resources. These options might include the type and number of compute nodes, the wall time limit, and your project account.

Single-node notebooks are launched in a dedicated queue, minimizing queueing time. For these notebooks, servers should be up and running within a few minutes. The maximum waiting time for a server to be running is 5 minutes, after which the job will be cancelled and you will be redirected back to the spawner options page. If your single-node server is not spawned within 5 minutes we encourage you to [contact us][ref-get-in-touch].

When resources are granted the page redirects to the JupyterLab session, where you can browse, open and execute notebooks on the compute nodes. A new notebook with a Python 3 kernel can be created with the menu `new` and then `Python 3` . Under `new` it is also possible to create new text files and folders, as well as to open a terminal session on the allocated compute node.

!!! tip "Debugging"

>     The log file of a JupyterLab server session is saved on `$HOME` in a file named `slurm-<jobid>.out`. If you encounter problems with your JupyterLab session, the contents of this file can be a good first clue to debug the issue.

Reviewer: ```suggestion
    The log file of a JupyterLab server session is saved on `$HOME` in a file named `slurm-<jobid>.out`. If you encounter problems with your JupyterLab session, the contents of this file can contain clues to debug the issue.
```

??? info "Using remote uenv for the first time."
    If the uenv is not present in the local repository, it will be automatically fetched.
    As a result, JupyterLab may take slightly longer than usual to start.

!!! warning "Ending your interactive session and logging out"
    The Jupyter servers can be shut down through the Hub. To end a JupyterLab session, please select `Hub Control Panel` under the `File` menu and then `Stop My Server`. By contrast, clicking `Logout` will log you out of the server, but the server will continue to run until the Slurm job reaches its maximum wall time.

> If the default base images do not meet your requirements, you can specify a custom environment instead. For this purpose, you supply either a custom uenv image/view or CE TOML file under the section `Advanced options` before launching the session. The supported uenvs are compatible with the Jupyter service out of the box, whereas container images typically require the installation of some additional packages. 

Reviewer: ```suggestion
If the default base images do not meet your requirements, you can specify a custom environment instead. For this purpose, you supply either a custom uenv image/view or [container engine (CE][ref-container-engine] TOML file under the section `Advanced options` before launching the session. The supported uenvs are compatible with the Jupyter service out of the box, whereas container images typically require the installation of some additional packages. 
```
This is the first time CE is mentioned, so I'd use the full name and a link (even if the same is already used for "container").

    RUN pip install --no-cache \
        jupyterlab \
        jupyterhub==4.1.6 \
        pyfirecrest==1.2.0 \
        SQLAlchemy==1.4.52 \
        oauthenticator==16.3.1 \
        notebook==7.3.3 \
        jupyterlab_nvdashboard==0.13.0 \
        git+https://github.com/eth-cscs/firecrestspawner.git
    ```

    The package [nvdashboard](https://github.com/rapidsai/jupyterlab-nvdashboard) is also installed here, which allows to monitor system metrics at runtime.
    
    A corresponding TOML file can look like

    ```toml
    image = "/capstor/scratch/cscs/${USER}/ce-images/ngc-pytorch+25.05.sqsh"

    mounts = [
        "/capstor", 
        "/iopsstor",
        "/users/${USER}/.local/share/jupyter", # (1)!
        "/etc/slurm", # (2)!
        "/usr/lib64/libslurm-uenv-mount.so",
        "/etc/container_engine_pyxis.conf" # (3)!
    ]

    workdir = "/capstor/scratch/cscs/${USER}" # (4)!

    writable = true

    [annotations]
    com.hooks.aws_ofi_nccl.enabled = "true" # (5)!
    com.hooks.aws_ofi_nccl.variant = "cuda12"

    [env]
    CUDA_CACHE_DISABLE = "1" # (6)!
    TORCH_NCCL_ASYNC_ERROR_HANDLING = "1" # (7)!
    MPICH_GPU_SUPPORT_ENABLED = "0" # (8)!
    ```
    

>     1. avoid mounting all of `$HOME` to avoid subtle issues with cached files, but mount Jupyter kernels
>     2. enable SLURM commands (together with two subsequent mounts)
>     3. currently only required on Daint and Santis, not on Clariden
>     4. set working directory of Jupyter session (file browser root directory)
>     5. use environment settings for optimized communication 
>     6. disable CUDA JIT cache
>     7. async error handling when an exception is observed in NCCL watchdog: aborting NCCL communicator and tearing down process upon error

Reviewer: ```suggestion
    1. Avoid mounting all of `$HOME` to avoid subtle issues with cached files, but mount Jupyter kernels
    2. Enable SLURM commands (together with two subsequent mounts)
    3. Currently only required on Daint and Santis, not on Clariden
    4. Set working directory of Jupyter session (file browser root directory)
    5. Use environment settings for optimized communication 
    6. Disable CUDA JIT cache
    7. Async error handling when an exception is observed in NCCL watchdog: aborting NCCL communicator and tearing down process upon error
```

As a preliminary step to running any code in Jupyter notebooks, a kernel needs to be installed, which is described in the following for both Python and Julia.

### Using Python in Jupyter

For Python, the recommended setup consists of a uenv or container as a base image as described [above][ref-jupyter-runtime-environment] that includes the stable dependencies of the software stack. Additional packages can be installed in a virtual environment _on top_ of the Python installation in the base image (mandatory for most uenvs). Having the base image loaded, such a virtual environment can be created with

```bash title="Create a virtual environment on top of a base image"
python -m venv --system-site-packages venv-<base-image-version>
```

> where `<base-image-version>` can be replaced by an identifier uniquely referring to the base image (such virtual environments are specific for the base image and non-portable).

Reviewer: ```suggestion
where `<base-image-version>` can be replaced by an identifier uniquely referring to the base image (such virtual environments are specific for the base image and are not portable).
```

!!! important "pass a [`julia`][ref-uenv-julia] uenv and the view `jupyter`."

When Julia is first used within Jupyter, IJulia and one or more Julia kernel need to be installed. 
Type the following command in a shell within JupyterHub to install IJulia, the default Julia kernel and, on systems whith Nvidia GPUs, a Julia kernel running under Nvidia Nsight Systems:
```bash
install_ijulia
```

You can install additional custom Julia kernels by typing the following in a shell:
```bash
julia
using IJulia
installkernel(<args>) # (1)!
```

1. type `? installkernel` to learn about valid `<args>`

!!! warning "First time use of Julia"
    If you are using Julia for the first time at all, executing `install_ijulia` will automatically first trigger the installation of `juliaup` and the latest `julia` version (it is also triggered if you execute `juliaup` or `julia`).

## Parallel computing

### MPI in the notebook via IPyParallel and MPI4Py

MPI for Python provides bindings of the Message Passing Interface (MPI) standard for Python, allowing any Python program to exploit multiple processors.

> MPI can be made available on Jupyter notebooks through [IPyParallel](https://github.com/ipython/ipyparallel). This is a Python package and collection of CLI scripts for controlling clusters for Jupyter: A set of servers that act as a cluster, called engines, is created and the code in the notebook's cells will be executed within them.

Reviewer: ```suggestion
MPI can be made available on Jupyter notebooks through [IPyParallel](https://github.com/ipython/ipyparallel). This is a Python package and collection of CLI scripts for controlling clusters for Jupyter: a set of servers that act as a cluster, called engines, is created and the code in the notebook's cells will be executed within them.
```

> We provide the python package [`ipcmagic`](https://github.com/eth-cscs/ipcluster_magic) to make easier the mangement of IPyParallel clusters. `ipcmagic` can be installed by the user with

Reviewer: ```suggestion
We provide the Python package [`ipcmagic`](https://github.com/eth-cscs/ipcluster_magic) to make easier the management of IPyParallel clusters. `ipcmagic` can be installed by the user with
```

Examples of notebooks with `ipcmagic` can be found [here](https://github.com/eth-cscs/ipcluster_magic/tree/master/examples).

### Distributed training and inference for ML

While it is generally recommended to submit long-running machine learning training and inference jobs via `sbatch`, certain use cases can benefit from an interactive Jupyter environment.

A popular approach to run multi-GPU ML workloads is with `accelerate` and `torchrun` as demonstrated in the [tutorials][ref-guides-mlp-tutorials]. In particular, the `accelerate launch` script in the [LLM fine-tuning tutorial][ref-mlp-llm-finetuning-tutorial] can be directly carried over to a Jupyter cell with a `%%bash` header (to run its contents interpreted by bash). For `torchrun`, one can adapt the command from the multi-node [nanotron tutorial][ref-mlp-llm-nanotron-tutorial] to run on a single GH200 node using the following line in a Jupyter cell

```bash
!torchrun --standalone --nproc_per_node=4 run_train.py ...
```

!!! warning "torchrun with virtual environments"

>     When using a virtual environment on top of a base image with Pytorch, replace `torchrun` with `python -m torch.distributed.run` to pick up the correct Python environment.

Reviewer: Can this be the default? It seems safer.

    
!!! note "Notebook structure"
    In none of these scenarios any significant memory allocations or background computations are performed on the main Jupyter process. Instead, the resources are kept available for the processes launched by `accelerate` or `torchrun`, respectively.

> Alternatively to using these launchers, it is also possible to use SLURM to obtain more control over resource mappings, e.g. by launching an overlapping SLURM step onto the same node used by the Jupyter process. An example with the container engine looks like this

Reviewer: ```suggestion
Alternatively to using these launchers, it is also possible to use SLURM to obtain more control over resource mappings, e.g. by launching an overlapping SLURM step onto the same node used by the Jupyter process. An example with the container engine looks like this:
```

By default, JupyterLab servers are launched in a dedicated queue, which should ensure a start-up time of less than a few minutes. If your server is not running within 5 minutes we encourage you to first try the non-dedicated queue, and then [contact us][ref-get-in-touch].

When resources are granted the page redirects to the JupyterLab session, where you can browse, open and execute notebooks on the compute nodes. A new notebook with a Python 3 kernel can be created with the menu `new` and then `Python 3` . Under `new` it is also possible to create new text files and folders, as well as to open a terminal session on the allocated compute node.

!!! tip "Debugging"
    The log file of a JupyterLab server session is saved in `$HOME` in a file named `slurm-<jobid>.out`. If you encounter problems with your JupyterLab session, the contents of this file can contain clues to debug the issue.

??? warning "Unexpected error while saving file: disk I/O error."
    This error message indicates that you have run out of disk quota.
    You can check your quota using the command `quota`.

> 

Reviewer 2: ```suggestion
```
